### PR TITLE
First implementation of getting MP3 from YouTube video and playing it…

### PR DESCRIPTION
**Pull request by joshvazquez**
…. #10

@nstbayless @wchae @sinoforever

### CODE REVIEW

File: routes/api.js [hunk 1]
```diff
@@ -76,17 +77,17 @@ module.exports = (upload) => {
     // sends back id of song
     function post_song_upload(req, res, next, list) {
         assert(!!req.file);
-        var parser = mm(fs.createReadStream(req.file.destination + req.file.filename), {duration: true}, function (err, metadata) {
+        var path = req.file.destination + req.file.filename;
+        var title = req.file.originalname;
+        var parser = mm(fs.createReadStream(path), {duration: true}, function (err, metadata) {
```
Comment (joshvazquez) [inline routes/api.js]: This parser section and the similar one in post_song_url() can be combined but I'm not sure how to deal with this given the callbacks.

File: routes/api.js [hunk 2]
```diff
@@ -149,6 +201,11 @@ module.exports = (upload) => {
         post(req, res, next);
     });
 
+    /* POST router, song url */
+    router.post(/.*\/songs\/url\/?$/, upload.single("song"), function (req, res, next) {
```
Comment (joshvazquez) [inline routes/api.js]: Not sure what to replace upload.single("song") with.

File: views/add.jade [hunk 3]
```diff
@@ -3,12 +3,15 @@ extends layout
...
-	div.q
-		h1 Add to #{pl.name}
-		p Only file upload supported currently
-		div.add-module-container
-			div
-				a(href="/add/#{pl.id}/upload")
-					img(src="/images/add_module_upload.png")
-				p Upload from your device
-			div(ng-repeat = "i in range(3)")
+    div.q
```
Comment (joshvazquez) [inline views/add.jade]: Replaced tabs and added button for URL. Needs better image.

**Comment by joshvazquez** [comment]
Also, this requires ffmpeg.

**Comment by nstbayless** [comment]
ffmpeg is fine. But I think it's been replaced by avconv

**Comment by nstbayless** [comment]
By the way, you might want to merge from master at some point. I keep seeing this error, for example, in these branches:

```
TypeError: /home/n/git/flyweb-mp3/views/layout.jade:24
    22| 			div#fab
    23| 				block fab
  > 24| 				a#fab-add(href="/add/#{pl.id}")
    25| 		footer
    26| 			block footer
    27| 			div#player-song
```

Merging from master fixed it (on my end, at least).

File: routes/api.js (hunk 2 above)
Comment (nstbayless) [inline routes/api.js]: That's from multer, for handling file uploads. It should be routed the other router.post route. I can take a look at that.

**Comment by nstbayless** [comment]
Okay, I tried it, fixed a merge error, it worked. Fantastic work!

The interface could probably be improved a bit (it does hang a while waiting to send a response to the POST request, for example). I can take a look at improving that, but it's a separate issue.

Merged into master.

**Comment by joshvazquez** [comment]
Thanks

**Pull request closed without merging**